File: go/src/common/wrapper.go
package common

import (
	"fmt"

	"github.com/jinzhu/copier"
)

type DataCarrier[P SourceData, T TargetData] struct {
	SourceData P
	TargetData T
}

type SourceData interface {
	DCDefault | interface{}
}

type DCDefault struct {
	Value  interface{}
	Assist interface{}
}

type TargetData interface {
	TDefault | interface{}
}

type TDefault struct {
	Value interface{}
}

type ProcessorMethod[P SourceData, T TargetData] func(P, T) bool

type Processor[P SourceData, T TargetData] struct {
	Fn ProcessorMethod[P, T]
}

type DataFlowProcessor[P SourceData, T TargetData] struct {
}

type DataFlowProcessorInterface[P SourceData, T TargetData] interface {
	ProcessData(processors []Processor[P, T], dataCarriers []DataCarrier[P, T]) interface{}
}

func (c DataFlowProcessor[P, T]) ProcessData(processors []Processor[P, T], dataCarriers []DataCarrier[P, T]) interface{} {

	channels := make(chan int)
	for index, processor := range processors {

		go func(processor Processor[P, T], carriers []DataCarrier[P, T], index int) {
			carrierCopy := DataCarrier[P, T]{}
			for _, carrier := range carriers {
				if err := copier.Copy(&carrierCopy, &carrier); err != nil {
					panic(err)
				}
				if !processor.Fn(carrierCopy.SourceData, carrierCopy.TargetData) {
					break
				}
			}

			channels <- index
		}(processor, dataCarriers, index)
	}

	for i := 0; i < len(processors); i++ {
		fmt.Println(<-channels)
	}

	return nil
}

File: rust/tests/common/wrapper.rs
#[derive(Debug)]
pub struct DataCarrier<S, T>
{
    pub source_data: S,
    pub target_data: T,
}

impl<S, T> DataCarrier<S, T>
{
    /// 构造函数
    pub fn new(source_data: S, target_data: T) -> Self {
        Self {
            source_data,
            target_data,
        }
    }
}

pub struct SourceData<V, A>
{
    pub value: V,
    pub assist: A,
}

impl<V, A> SourceData<V, A>
{
    /// 构造函数
    pub fn new(value: V, assist: A) -> Self {
        Self {
            value,
            assist,
        }
    }
}

pub struct TargetData<V>
{
    pub value: V,
}

impl<V> TargetData<V>
{
    /// 构造函数
    pub fn new(value: V) -> Self {
        Self {
            value,
        }
    }
}


